virttest/utils_test/libvirt.py

```diff
@@ -1438,47 +1438,55 @@ def create_hostdev_xml(pci_id, boot_order=None, xmlfile=True,
-    Create controller xml or add controller
+    Add a specified controller to the vm xml
+    :param vm_name: The vm name to be added with a controller
+    :param contr: The controller object to be added
+    :return: The newly added controller object
+    """
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    org_controllers = vmxml.get_controllers(contr.type)
+    vmxml.add_device(contr)
+    vmxml.sync()
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    updated_controllers = vmxml.get_controllers(contr.type)
+    for contr in updated_controllers:
+        if contr not in org_controllers:
+            new_added_ctr = contr
+    if 'new_added_ctr' not in locals():
+        raise exceptions.TestFail("Fail to get new added controller.")
+    return new_added_ctr
+
+
+def create_controller_xml(contr_dict):
+    """
+    Create a controller xml
 
     :param contr_dict: The dict params includs controller configurations
-    :param oper: Operation need to do, current we have 2 choice
-                1. "get_xml": controller xml according to dict
-                2. "add_controller": add controller according to dict and
-                updated in vm
-    :param vm_name: name of vm, mandatory for 'add_controller' operation
-    :return: controller xml file if oper is equal to "get_xml";
-             controller object for 'add_controller'.
+    :return: a tuple of (new controller's xml, new controller object)
     """
     contr_type = contr_dict.get("controller_type", 'scsi')
     contr_model = contr_dict.get("controller_model", "virtio-scsi")
     contr_index = contr_dict.get("controller_index")
     contr_alias = contr_dict.get("contr_alias")
     contr_addr = contr_dict.get("controller_addr")
+    contr_busNr = contr_dict.get("controller_busNr")
+    contr_node = contr_dict.get("controller_node")
     contr = controller.Controller(contr_type)
     contr.model = contr_model
     if contr_index:
         contr.index = contr_index
     if contr_alias:
         contr.alias = dict(name=contr_alias)
+    if contr_busNr:
+        contr.target = dict(busNr=contr_busNr)
+    if contr_node:
+        contr.node = contr_node
     if contr_addr:
         contr.address = contr.new_controller_address(attrs=eval(contr_addr))
-    if oper == "add_controller":
-        vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
-        org_controllers = vmxml.get_controllers(contr_type)
-        vmxml.add_device(contr)
-        vmxml.sync()
-        vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
-        updated_controllers = vmxml.get_controllers(contr_type)
-        for contr in updated_controllers:
-            if contr not in org_controllers:
-                new_added_ctr = contr
-        if 'new_added_ctr' not in locals():
-            raise exceptions.TestFail("Fail to get new added controller.")
-        return new_added_ctr
-    elif oper == "get_xml":
-        return contr.xml
+
+    return (contr.xml, contr)
```

Comment:
I assume returning controller object only is enough since contr.xml is one attribute of controller object

Reply:
update